libopenage/unit/action.cpp

```diff
@@ -1000,6 +1009,52 @@ void AttackAction::fire_projectile(const Attribute<attr_type::attack> &att, cons
 }
 
 
+HealAction::HealAction(Unit *e, UnitReference tar)
+	:
+	TargetAction{e, graphic_type::heal, tar, get_attack_range(e)},
+	heal_percent{0.0f},
+	rate_of_heal{0.004f} {
+
+}
+
+HealAction::~HealAction() {}
+
+void HealAction::update_in_range(unsigned int time, Unit *target_ptr) {
+	if (this->heal_percent > 0.0) {
+		this->heal_percent -= this->rate_of_heal * time;
+	}
+	else {
+		this->heal_percent += 1.0f;
+		this->heal(*target_ptr);
+	}
+
+	// inc frame
+	this->frame += time * this->current_graphics().at(graphic)->frame_count * this->rate_of_heal;
+}
+
+bool HealAction::completed_in_range(Unit *target_ptr) const {
+	auto &h_attr = target_ptr->get_attribute<attr_type::hitpoints>();
+	return h_attr.current >= h_attr.max; // is unit at full hitpoints?
```

Comment:
At source: `int current; // can become negative`, so I belive it may break things.
Also I am not fully understand what "extend the value fetching" means...